packages/frontend/hooks/src/use-app-updater.ts

```diff
@@ -0,0 +1,233 @@
+  return rpcToObservable(null as number | null, {
+    event: window.events?.updater.onDownloadProgress,
+  });
+});
+
+export const changelogCheckedAtom = atomWithStorage<Record<string, boolean>>(
+  'affine:client-changelog-checked',
+  {}
+);
+
+export const currentVersionAtom = atom(async () => {
+  if (!isBrowser) {
+    return null;
+  }
+  const currentVersion = await window.apis?.updater.currentVersion();
+  return currentVersion;
+});
+
+export const currentChangelogUnreadAtom = atom(async get => {
+  if (!isBrowser) {
+    return false;
+  }
+  const mapping = get(changelogCheckedAtom);
+  const currentVersion = await get(currentVersionAtom);
+  if (currentVersion) {
+    return !mapping[currentVersion];
+  }
+  return false;
+});
+
+export const isCheckingForUpdatesAtom = atom(false);
+
+export const isAutoDownloadUpdateAtom = atom(async () => {
+  try {
+    const res = await window.apis?.updater.autoDownloadUpdate();
+    return res;
+  } catch (err) {
+    console.error(err);
+    return true;
+  }
+});
+
+export const isAutoCheckUpdateAtom = atom(async () => {
+  try {
+    const res = await window.apis?.updater.autoCheckUpdate();
+    return res;
+  } catch (err) {
+    console.error(err);
+    return true;
+  }
+});
+
+export const useAppUpdater = () => {
+  const currentChangelogUnread = useAtomValue(currentChangelogUnreadAtom);
+  const [appQuitting, setAppQuitting] = useState(false);
+  const setChangelogCheckAtom = useSetAtom(changelogCheckedAtom);
+  const updateReady = useAtomValue(updateReadyAtom);
+  const updateAvailable = useAtomValue(updateAvailableAtom);
+  const currentVersion = useAtomValue(currentVersionAtom);
+  const setUpdateAvailableState = useSetAtom(updateAvailableStateAtom);
+  const setIsCheckingForUpdates = useSetAtom(isCheckingForUpdatesAtom);
+
+  const dismissCurrentChangelog = useCallback(() => {
+    if (!currentVersion) {
+      return;
+    }
+    startTransition(() =>
+      setChangelogCheckAtom(mapping => {
+        return {
+          ...mapping,
+          [currentVersion]: true,
+        };
+      })
+    );
+  }, [currentVersion, setChangelogCheckAtom]);
+
+  const quitAndInstall = useCallback(() => {
+    if (updateReady) {
+      setAppQuitting(true);
+      window.apis?.updater.quitAndInstall().catch(err => {
+        // TODO: add error toast here
+        console.error(err);
+      });
+    } else if (updateAvailable) {
+      if (updateAvailable.allowAutoUpdate) {
+        // wait for download to finish
+      } else {
+        window.open(
+          `https://github.com/toeverything/AFFiNE/releases/tag/v${currentVersion}`,
+          '_blank'
+        );
+      }
+    } else if (currentChangelogUnread) {
+      window.open(runtimeConfig.changelogUrl, '_blank');
+      dismissCurrentChangelog();
+    } else {
+      throw new Unreachable();
+    }
+  }, [
+    currentChangelogUnread,
+    currentVersion,
+    dismissCurrentChangelog,
+    updateAvailable,
+    updateReady,
+  ]);
```

Comment:
Implementations of `dismissCurrentChangelog` and `quitAndInstall` etc were for the app updater button only. It shall be refactored for general purpose if you want to move out of it.